File: windows/tray-app/RifeZPhoneBridge.Core/Protocol/FlatBufferFrameIO.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using System.Buffers.Binary;

namespace RifeZPhoneBridge.Core.Protocol;

public static class FlatBufferFrameIO
{
    public static async Task WriteFrameAsync(Stream stream, byte[] payload, CancellationToken cancellationToken = default)
    {
        byte[] header = new byte[4];
        BinaryPrimitives.WriteInt32LittleEndian(header, payload.Length);

        await stream.WriteAsync(header, cancellationToken);
        await stream.WriteAsync(payload, cancellationToken);
        await stream.FlushAsync(cancellationToken);
    }

    public static async Task<byte[]?> ReadFrameAsync(Stream stream, CancellationToken cancellationToken = default)
    {
        byte[] header = new byte[4];
        int headerRead = await ReadExactlyAsync(stream, header, 0, 4, cancellationToken);
        if (headerRead == 0)
            return null;

        if (headerRead < 4)
            throw new IOException("Incomplete FlatBuffer frame header.");

        int length = BinaryPrimitives.ReadInt32LittleEndian(header);
        if (length <= 0)
            throw new IOException($"Invalid FlatBuffer frame length: {length}");

        byte[] payload = new byte[length];
        int payloadRead = await ReadExactlyAsync(stream, payload, 0, length, cancellationToken);
        if (payloadRead < length)
            throw new IOException("Incomplete FlatBuffer frame payload.");

        return payload;
    }

    private static async Task<int> ReadExactlyAsync(
        Stream stream,
        byte[] buffer,
        int offset,
        int count,
        CancellationToken cancellationToken)
    {
        int totalRead = 0;

        while (totalRead < count)
        {
            int read = await stream.ReadAsync(
                buffer.AsMemory(offset + totalRead, count - totalRead),
                cancellationToken);

            if (read == 0)
                return totalRead;

            totalRead += read;
        }

        return totalRead;
    }
}

File: android/receiver-app/app/src/main/java/com/rifez/phoneaudio/protocol/FlatBufferFrameIO.kt
package com.rifez.phoneaudio.protocol

import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

object FlatBufferFrameIO {

    fun writeFrame(outputStream: OutputStream, payload: ByteArray) {
        val header = ByteBuffer.allocate(4)
            .order(ByteOrder.LITTLE_ENDIAN)
            .putInt(payload.size)
            .array()

        outputStream.write(header)
        outputStream.write(payload)
        outputStream.flush()
    }

    fun readFrame(inputStream: InputStream): ByteArray? {
        val header = ByteArray(4)
        val headerRead = readExactly(inputStream, header, 4)
        if (headerRead == 0) return null
        if (headerRead < 4) throw IllegalStateException("Incomplete FlatBuffer frame header")

        val length = ByteBuffer.wrap(header)
            .order(ByteOrder.LITTLE_ENDIAN)
            .int

        require(length > 0) { "Invalid FlatBuffer frame length: $length" }

        val payload = ByteArray(length)
        val payloadRead = readExactly(inputStream, payload, length)
        if (payloadRead < length) {
            throw IllegalStateException("Incomplete FlatBuffer frame payload")
        }

        return payload
    }

    private fun readExactly(inputStream: InputStream, buffer: ByteArray, count: Int): Int {
        var totalRead = 0
        while (totalRead < count) {
            val read = inputStream.read(buffer, totalRead, count - totalRead)
            if (read == -1) return totalRead
            totalRead += read
        }
        return totalRead
    }
}